transaction/std/src/memo_builder/gift_code_cancellation_memo_builder.rs

```diff
@@ -57,53 +73,106 @@ impl MemoBuilder for GiftCodeCancellationMemoBuilder {
-        let memo_payload = build_zero_value_change_memo(&mut builder).unwrap();
+        let memo_payload = build_change_memo_with_amount(&mut builder, amount).unwrap();
         let memo_data = memo_payload.get_memo_data();
 
-        // Verify memo data
-        let derived_index = u64::from_le_bytes(memo_data[0..8].try_into().unwrap());
-        assert_eq!(index, derived_index);
+        // Recover memo
+        let memo = GiftCodeCancellationMemo::from(memo_data);
+
+        // Check recovered index is correct
+        assert_eq!(index, memo.cancelled_gift_code_index());
+
+        // Check recovered fee is correct
+        assert_eq!(fee.value, memo.get_fee());
     }
 
     #[test]
     fn test_gift_code_cancellation_memo_fails_for_more_than_one_change_memo() {
-        // Set the cancellation index and create memo builder
+        // Set the cancellation index and fee and create the memo builder
         let index = 666;
+        let amount = Amount::new(42, 0.into());
+        let fee = Amount::new(20, 0.into());
         let mut builder = GiftCodeCancellationMemoBuilder::new(index);
+        builder.set_fee(fee).unwrap();
 
         // Attempt to build two change outputs
-        build_zero_value_change_memo(&mut builder).unwrap();
-        let memo_payload = build_zero_value_change_memo(&mut builder);
-
-        // Assert failure for the second output
-        assert!(matches!(
-            memo_payload,
-            Err(NewMemoError::MultipleChangeOutputs)
-        ));
+        let memo_payload = build_change_memo_with_amount(&mut builder, amount).unwrap();
+        let memo_payload_2 = build_change_memo_with_amount(&mut builder, amount);
+
+        // Ensure we can recover the index and fee for the first change output
+        let memo = GiftCodeCancellationMemo::from(memo_payload.get_memo_data());
+        assert_eq!(index, memo.cancelled_gift_code_index());
+        assert_eq!(fee.value, memo.get_fee());
+
+        // Assert failure for the second attempted change output
+        assert_eq!(memo_payload_2, Err(NewMemoError::MultipleChangeOutputs));
+    }
+
+    #[test]
+    fn test_gift_code_cancellation_memo_builder_fee_token_cannot_be_different_from_change_token() {
+        let index = 666;
+        let mut builder = GiftCodeCancellationMemoBuilder::new(index);
+        let amount = Amount::new(42, 0.into());
+        let fee = Amount::new(1, 9001.into());
+
+        // Set a fee with a different token id
+        builder.set_fee(fee).unwrap();
+
+        // Attempt to build the memo
+        let memo_payload = build_change_memo_with_amount(&mut builder, amount);
+
+        // Ensure memo creation fails
+        assert_eq!(memo_payload, Err(NewMemoError::MixedTokenIds));
+    }
+
+    #[test]
+    fn test_gift_code_cancellation_memo_builder_set_fee_fails_when_exceeding_max_fee() {
+        let index = 666;
+        let mut builder = GiftCodeCancellationMemoBuilder::new(index);
+        let fee = Amount::new(u64::MAX, 0.into());
```

Comment:
This should be testing boundary conditions:
* `GiftCodeCancellationMemo::MAX_FEE`
* `GiftCodeCancellationMemo::MAX_FEE + 1`